packages/react/types/index.d.ts

```diff
@@ -46,6 +46,24 @@ export function withEmotionCache<Props, RefType = any>(
   func: (props: Props, context: EmotionCache, ref: Ref<RefType>) => ReactNode
 ): FC<Props & ClassAttributes<RefType>>
 
+type ReactInterpolationPrimitive =
+  | null
+  | undefined
+  | boolean
+  | SerializedStyles
+  | CSSObject
+
+type ReactArrayInterpolation<Props> = Array<ReactInterpolation<Props>>
+
+interface ReactFunctionInterpolation<Props> {
+  (props: Props): ReactInterpolation<Props>
+}
+
+export type ReactInterpolation<Props> =
+  | ReactInterpolationPrimitive
+  | ReactArrayInterpolation<Props>
+  | ReactFunctionInterpolation<Props>
+
```

Comment:
Can we call the type parameter `Theme` since that's what it is here?

Reply:
Ive also used it for one of the styled overloads where it is equivalent to Props

Reply:
This type doesn't seem right for styled, e.g. `styled.div('display:flex;')` is valid.

Also, this change isn't right for the css prop either since `css={['display:flex;']}` is valid. The only thing that's wrong about the current types is that `string` isn't allowed at the root of the css prop.

Reply:
Well, sure - the question is if we want to support this. Types don't always have to match the runtime reality fully. It feels better to guide users to better patterns if they are using a type checker.

Reply:
@mitchellhamilton oh ok, I must have missed these style string patterns in the doc because I based my suggestion on what I saw demonstrated. If we really want to allow this form of string we can go ultra geeky and use template literal types but it seems overengineering at this point.

To be honest, the one thing I'm really interested in is for TS to catch me (and anyone starting with the `css` property) when I'm worngly assigning a class name to the `css` property.

So I'd agree with @Andarist here. It's about nudging and guiding to use some patterns rather that fully typing everything that could work but it's up to you folks.

Reply:
Sure, let's change this for the css prop. Can you revert the change for @emotion/styled though since having @emotion/styled depend on a new thing from @emotion/react would be a breaking change since @emotion/styled has a peer dependency on @emotion/react? (also for styled, i believe we compile styles to `styled.div('display: ', something, ';')` and making it so our types treat something that we compile to as invalid seems bad)

Reply:
Ok, I've removed this from `@emotion/styled` and also renamed Props to Theme since this will always be true right now.

This might be a user-facing change (some people might use `Interpolation` type in their code) but I think this is OK if we make this a minor change. Tying major versioning to types (unless changes are really extreme) falls outside of the pragmatic semver for me. What's your opinion on this?